src/libraries/System.Net.Http/src/System/Net/Http/MetricsHandler.cs

```diff
@@ -0,0 +1,252 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Net.Http.Metrics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http;
+
+internal sealed class MetricsHandler : HttpMessageHandlerStage, IHttpMetricsLogger
+{
+    private readonly HttpMessageHandler _innerHandler;
+    private readonly UpDownCounter<long> _currentRequests;
+    private readonly Counter<long> _failedRequests;
+    private readonly Histogram<double> _requestsDuration;
+
+    private ThreadLocal<HttpMetricsEnrichmentContext> _cachedEnrichmentContext = new ThreadLocal<HttpMetricsEnrichmentContext>(() => new HttpMetricsEnrichmentContext());
```

Comment:
I haven't used `ThreadLocal<T>` much. Should it be static?

I still wonder whether the overhead of pooling is worth saving a couple of small allocations per HTTP request.

Reply:
Yeah it should probably be static.

> I still wonder whether the overhead of pooling is worth saving a couple of small allocations per HTTP request.

There are probably perf experts with more experience making these calls, but this is how I think of it:
- If a code path allocates < 100KB/sec under a high load app then we are probably better off optimizing for code simplicity. I think this code path is hot enough that we aren't in that case.
- In favorable circumstances an allocation of a minimum sized object + amortized GC time is in the ballpark of 10ns. In practice the actual times depend on a bunch of factors, most of which will cause realistic app conditions to be same or worse as microbenchmark performance.
- Aside from impacts to throughput, there are also impacts to tail latency that are hard to measure with simple benchmarks but potentially more important to services than raw throughput. In general more allocation => more frequent GCs => more likely an object of a given lifetime survives into higher generations => more expensive GCs. So even if throughput differences were a wash, latency and tail latency improvements can still make reduced allocation a win.

This code is going much further than I was expecting trying to guard against mis-use of the pooled objects. IMO just clearing the fields of the object is a sufficient guard because anyone who tries to use the object later won't get useful values from it. I was imagining the complexity cost of pooling being < 10 lines of code.

Reply:
Yeah it should probably be static.

> I still wonder whether the overhead of pooling is worth saving a couple of small allocations per HTTP request.

There are probably perf experts with more experience making these calls, but this is how I think of it:
- If a code path allocates < 1000 objects/sec/core under high load then we are probably better off optimizing for code simplicity. I don't think that is our case here. I assume high scale services could easily hit ~10,000 objects/sec/core on this path bec

- In very favorable circumstances an allocation of a minimum sized object + amortized GC time is ~10ns. In practice the actual times depend on object size, CPU cache size, memory bus bandwidth, locality of data access, survival rates, GC heap size, etc. Nearly all of those factors represent risks that the allocation will be slower under more realistic app conditions vs. microbenchmark conditions. I expect its pretty easy for a threadlocal singleton cache to pay for itself on throughput for any code path you expect to be somewhat hot.
- The more rapidly code allocates the more frequently GC will run. This in turn makes it more likely that short lived objects get promoted into higher generations

Reply:
> Yeah it should probably be static.

That's `[ThreadStatic]`. I wanted these context instances to be bound not only the thread, but also the `MetricsHandler` instance, but realized that it's unnecessary, so I will switch to `[ThreadStatic]`.

Reply:
> I still wonder whether the overhead of pooling is worth saving a couple of small allocations per HTTP request.

We have fought and clawed our way to better perf with HttpClient, in large part one allocation at a time. Here we're talking about potentially hundreds of bytes of allocation per request. We should either avoid the allocation entirely (stackalloc the necessary state and pass it by ref) or find a way to pool it cheaply. A single TLS hit is as cheap as it gets. But avoiding the allocation entirely is still my preference if possible; I'm not worried about minor usability impact for an API only a handful of devs will ever use directly. There are other more subtle costs that come with pooling, eg every string/object that gets stored in the tags will end up making gen0 gcs more expensive due to the gen2 to gen0 references.